File: src/services/signupService/signupService.go
package signupService

import (
	"context"
	"fmt"
	"i9rfs/src/appTypes"
	"i9rfs/src/helpers"
	user "i9rfs/src/models/userModel"
	"i9rfs/src/services/mailService"
	"i9rfs/src/services/securityServices"
	"os"
	"time"

	"github.com/gofiber/fiber/v2"
)

func RequestNewAccount(ctx context.Context, email string) (any, map[string]any, error) {

	userExists, err := user.Exists(ctx, email)
	if err != nil {
		return nil, nil, err
	}

	if userExists {
		return nil, nil, fiber.NewError(fiber.StatusBadRequest, "A user with this email already exists.")
	}

	verfCode, expires := securityServices.GetTokenAndExpiration()

	go mailService.SendMail(email, "Verify your email", fmt.Sprintf("<p>Your email verification code is <strong>%s</strong></p>", verfCode))

	sessionData := map[string]any{
		"email":        email,
		"vCode":        verfCode,
		"vCodeExpires": expires,
	}

	respData := map[string]any{
		"msg": fmt.Sprintf("Enter the 6-digit code sent to %s to verify your email", email),
	}

	return respData, sessionData, nil
}

func VerifyEmail(ctx context.Context, sessionData map[string]any, inputVerfCode string) (any, map[string]any, error) {
	var sd struct {
		Email        string
		VCode        string
		VCodeExpires time.Time
	}

	helpers.ToStruct(sessionData, &sd)

	if sd.VCode != inputVerfCode {
		return nil, nil, fiber.NewError(fiber.StatusBadRequest, "Incorrect verification code! Check or Re-submit your email.")
	}

	if sd.VCodeExpires.Before(time.Now()) {
		return nil, nil, fiber.NewError(fiber.StatusBadRequest, "Verification code expired! Re-submit your email.")
	}

	go mailService.SendMail(sd.Email, "Email Verification Success", fmt.Sprintf("Your email <strong>%s</strong> has been verified!", sd.Email))

	newSessionData := map[string]any{"email": sd.Email}

	respData := map[string]any{
		"msg": fmt.Sprintf("Your email, %s, has been verified!", sd.Email),
	}

	return respData, newSessionData, nil
}

func RegisterUser(ctx context.Context, sessionData map[string]any, username, password string) (any, string, error) {
	email := sessionData["email"].(string)

	userExists, err := user.Exists(ctx, username)
	if err != nil {
		return nil, "", err
	}

	if userExists {
		return nil, "", fiber.NewError(fiber.StatusBadRequest, "Username not available")
	}

	hashedPassword, err := securityServices.HashPassword(password)
	if err != nil {
		return nil, "", err
	}

	newUser, err := user.New(ctx, email, username, hashedPassword)
	if err != nil {
		return nil, "", err
	}

	authJwt, err := securityServices.JwtSign(appTypes.ClientUser{
		Username: username,
	}, os.Getenv("AUTH_JWT_SECRET"), time.Now().UTC().Add(10*24*time.Hour)) // 10 days

	if err != nil {
		return nil, "", err
	}

	respData := map[string]any{
		"msg":  "Signup success!",
		"user": newUser,
	}

	return respData, authJwt, nil
}

File: src/services/appServices/signupService.ts
import { StatusCodes } from "http-status-codes"
import * as user from "../../models/userModel.ts"
import * as securityServices from "./securityServices.ts"
import * as mailService from "../otherServices/mailService.ts"
import type { ClientUserT } from "../../appTypes.ts"

interface SignupRNASessionData {
  email: string
  vCode: string
  vCodeExpires: number
}

interface SignupRNAResp {
  msg: string
}

export async function RequestNewAccount(email: string) {
  if (await user.Exists(email)) {
    throw {
      name: "AppError",
      code: StatusCodes.BAD_REQUEST,
      message: "A user with this email already exists.",
    }
  }

  const { token: verfCode, expires } = securityServices.GetTokenAndExpiration()

  mailService.SendMail(
    email,
    "Verify your email",
    `<p>Your email verification code is <strong>${verfCode}</strong></p>`
  )

  const sessionData: SignupRNASessionData = {
    email: email,
    vCode: verfCode,
    vCodeExpires: expires,
  }

  const respData: SignupRNAResp = {
    msg: `Enter the 6-digit code sent to ${email} to verify your email`,
  }

  return { respData, sessionData }
}

interface SignupVESessionData {
  email: string
}

interface SignupVEResp {
  msg: string
}

export async function VerifyEmail(
  sd: SignupRNASessionData,
  inputVerfCode: string
) {
  if (sd.vCode != inputVerfCode) {
    throw {
      name: "AppError",
      code: StatusCodes.BAD_REQUEST,
      message: "Incorrect verification code! Check or Re-submit your email.",
    }
  }

  if (sd.vCodeExpires < Date.now()) {
    throw {
      name: "AppError",
      code: StatusCodes.BAD_REQUEST,
      message: "Verification code expired! Re-submit your email.",
    }
  }

  mailService.SendMail(
    sd.email,
    "Email Verification Success",
    `Your email <strong>${sd.email}</strong> has been verified!`
  )

  const newSessionData: SignupVESessionData = { email: sd.email }

  const respData: SignupVEResp = {
    msg: `Your email, ${sd.email}, has been verified!`,
  }

  return { respData, newSessionData }
}

interface SignupRUResp {
  msg: string
  user: ClientUserT
}

export async function RegisterUser(
  sessionData: SignupVESessionData,
  username: string,
  password: string
) {
  const { email } = sessionData

  if (await user.Exists(username)) {
    throw {
      name: "AppError",
      code: StatusCodes.BAD_REQUEST,
      message: "Username not available",
    }
  }

  const newUser = await user.New(
    email,
    username,
    await securityServices.HashPassword(password)
  )

  const authJwt = securityServices.JwtSign(
    newUser,
    process.env.AUTH_JWT_SECRET || "",
    10 * 24 * 60 * 60 * 1000
  ) // 10 days

  const respData: SignupRUResp = {
    msg: "Signup success!",
    user: newUser,
  }

  return { respData, authJwt }
}
